### PR TITLE
Issue #2329: Updated lexer grammar to allow javadocs with CR newlines

### PR DESCRIPTION
**Pull request by ps-sp**
#2329: Updated lexer grammar to allow javadocs with CR newlines

### CODE REVIEW

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4 [hunk 1]
```diff
@@ -37,8 +37,7 @@ import java.util.*;
 
 }
 
-LEADING_ASTERISK : ( (' '|'\t') {_tokenStartCharPositionInLine == 0}? ) (' '|'\t')* '*'
-      | '*' {_tokenStartCharPositionInLine == 0}?
+LEADING_ASTERISK : (' '|'\t')* '*'
```
Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: Please explain this change.

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: @rnveach Actaully it wasn't recognizing leading tokens after a newline so I had two options either to modify the entire rule or add the alternative `| '*' {previousTokenType = NEWLINE}?`

I chose the former and just tried to simplify the token def as much as I could based on the basic format of javadoc. Since the UTs seem to be insufficient, I was hoping to realize if there is any problem with this update, with the regression testing tool which is to be developed without which this PR isn't getting merged anyway.

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: Perhaps if I just added the alternative then regression testing would not have been necessary. Anyway, I felt it was better to simplify things as much as I could.

Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: > it wasn't recognizing leading asterisks after a newline 

You mean outside the `\r` new line change? If so, can you make a report to show this issue?

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: no only when the newline begins with '\r'

Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: I didn't realize `_tokenStartCharPositionInLine` is a field that came from ANTLR itself.
It is populated via `_tokenStartCharPositionInLine = getInterpreter().getCharPositionInLine();` which is defined at https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/LexerATNSimulator.java#L736-L746 .
It is only reset to 0 on `\n` so this is why the fix with `\r` doesn't work anymore. It doesn't see it as a new line.

I question why we had this special code before if it was never really needed.
@ps-sp If you print the ANTLR tree with only `\r` does it display the line numbers the same as if you were using `\n`?

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: @rnveach Yes and thus I had to choose from one of the two strategies I mentioned before. I was also wondering that where exactly would my token def and the present token def would create the differences but couldn't find anything and concluded that I shall make a PR and find if there are going to be problems in regression testing.

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: JAVADOC -> * Test case for detecting CR newlines.\r* @author Author Name\r<EOF> [0:0]
|--LEADING_ASTERISK -> * [0:0]
|--TEXT ->  Test case for detecting CR newlines. [0:1]
|--NEWLINE -> \r [0:38]
|--LEADING_ASTERISK -> * [0:39]
|--WS ->   [0:40]
|--JAVADOC_TAG -> @author Author Name\r [0:41]
|   |--AUTHOR_LITERAL -> @author [0:41]
|   |--WS ->   [0:48]
|   `--DESCRIPTION -> Author Name\r [0:49]
|       |--TEXT -> Author Name [0:49]
|       `--NEWLINE -> \r [0:60]
`--EOF -> <EOF> [0:61]

Here is the tree, and yes it fails to recognize new lines !

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: Should we create an issue for ANTLR ?

Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: @ps-sp Your tests also fail to display any line/column numbers as I can't find any of your print out numbers in it.

> Should we create an issue for ANTLR project ?

Yes, please create an issue with them.

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: @rnveach Actually the tests are spit out by [this](https://github.com/checkstyle/contribution/blob/master/javadoc-expected-tree-generator/ExpectedParseTreeGenerator.java) in contribution as you might know. So I was just sticking to it. All this should be resolved in [4381](https://github.com/checkstyle/checkstyle/issues/4381).

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: [Issue](https://github.com/antlr/antlr4/issues/1890) created for antlr4.

Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: @romani Please weigh in on this issue. Antlr has said it will never support `\r` alone as a new line.
They recommend converting `\r` alone to `\n` before parsing.

Should we do this?

Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: no, we should not do any conversions to sources, at least till clearly reported as bug with many supporters.
So for now lets keep this is as limitation.

**Comment by rnveach** [comment]
@ps-sp Here is the result of a regression with this change.
http://rveach.no-ip.org/checkstyle/regression/reports/64/

Looking over the reports, the changes at http://rveach.no-ip.org:9998/reports/64/checkstyle/index.html#A3 , seem incorrect to me.
Code: https://github.com/checkstyle/checkstyle/blob/25a37e5049b7816b34c552899841a978efc37a63/src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule231filetab/InputFileTabCharacter.java#L11
The final `*` is now being considered a leading asterisk, but I believe leading asterisks must start the line, which is why I think this change is wrong.
@ps-sp Please confirm the changes in the report.
@romani Please confirm I am right that these changes aren't warranted.

-----

Since we keep going over this new regression, I have been diving into it more and created my own process yesterday.
I created the report with the combination of https://github.com/rnveach/contribution/commits/text_comparing (for patch-diff-report-util) and https://github.com/rnveach/contribution/commits/rvmaster (launch_diff_antlr.sh).
It is almost a copy/paste of my other script, `launch_diff.sh` which was turned into the groovy scripts. It just takes the PR branch as a name and does everything else for you automatically.
Sadly it took more than 8 hours to do all these projects, so I really recommend small projects. I will look into trying to multi-thread the utility.
`patch-diff-report-util` can't handle xrefs to non-java files, so if we go through with this process, I will make an issue and fix it.

**Comment by ps-sp** [comment]
@rnveach The second approach is safer and I think that's almost certain to work: 
     
    LEADING_ASTERISK : ( (' '|'\t') {_tokenStartCharPositionInLine == 0}? ) (' '|'\t')* '*'
           | '*' {_tokenStartCharPositionInLine == 0}?
           | '*' {previousTokenType == NEWLINE}?
           ;

**Comment by Vladlis** [comment]
@ps-sp 
Please, prove it with UTs

**Comment by ps-sp** [comment]
@Vladlis The UT for this particular issue already exists. It works with the proposed change. I think for now we can rely only on regression testing.

**Comment by rnveach** [comment]
@ps-sp Your UT doesn't have asterisks in the middle, which is what regression has shown us is a problem with your change. Your UT's asterisk is only at start.
Here is another example from regression of any asterisk being changed to leading asterisk: http://rveach.no-ip.org/checkstyle/regression/reports/64/lombok-ast/index.html#A1

Please add this as a new UT to show original behavior, before this PR, and that your proposed updates won't break it. This will be a new quick regression, and once we can duplicate it, than we can re-run regression to see if there are anymore issues.

**Comment by rnveach** [comment]
@ps-sp Minor change was merged, so please rebase this PR.
Please verify test passes with your CR grammar changes.

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/javadocTags/InputJavadocWithCrAsNewline.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+* Test case for detecting CR newlines.* @author Author Name
```
Comment (Vladlis) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/javadocTags/InputJavadocWithCrAsNewline.txt]: Please add inputs with whitespaces and tabs after `\r` and before `*`

Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/javadocTags/InputJavadocWithCrAsNewline.txt]: Yes, it looks like you may have to switch all `_tokenStartCharPositionInLine == 0` to `_tokenStartCharPositionInLine == 0 || previousTokenType == NEWLINE`.

Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/javadocTags/InputJavadocWithCrAsNewline.txt]: @ps-sp It is probably best for your UT that you make an exact duplicate of `InputLeadingAsterisksExtended.txt` but change all `\n` to `\r`.

**Comment by rnveach** [comment]
@ps-sp As you make the changes and find a working solution, please also re-run regression with our beta ANTLR regression script found at https://github.com/checkstyle/contribution/issues/237#issuecomment-307267485 .

**Comment by rnveach** [comment]
@ps-sp ping

**Comment by ps-sp** [comment]
@rnveach made a few changes. Please have a look. Will generate regression reports finally. Regression takes a lot of time on my system, and on top of it hangs up sometimes while checking some large file.

**Comment by rnveach** [comment]
> Regression takes a lot of time on my system, and on top of it hangs up sometimes while checking some large file.

I'm not sure if your mentor has a strict order of issues, but maybe you should put optimizing the ANTLR code as a higher priority as you will have to deal with the slowness until than.

File: src/test/java/com/puppycrawl/tools/checkstyle/BaseAstTestSupport.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class BaseAstTestSupport {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/BaseAstTestSupport.java]: 2nd PR I have seen creating this new file. I had other review items on this class in another PR.

If we really want this class' methods in, why don't we do #4381 which could be merged first as it is just a test change and won't need regression.

**Comment by ps-sp** [comment]
@rnveach I am not trying to optimize regression testing. Just trying to run the script as it is again and again until it works. It pauses endlessly on random files. And yes all the antlr issues are prioritized. There is an order. Also there seems to be issue with `NewlineAtEndOfFile` verifying the input file with Cr as new lines. It fails to recognize the newline at end of file which should be present as even the expectedAst suggests.

File: src/test/java/com/puppycrawl/tools/checkstyle/BaseAstTestSupport.java (hunk 3 above)
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/BaseAstTestSupport.java]: Discussion made at https://groups.google.com/forum/?hl=en#!topic/checkstyle-devel/cRw80h4Q73o so it is in a clear place instead of spread among many PRs

**Comment by rnveach** [comment]
> there seems to be issue with NewlineAtEndOfFile verifying the input file with Cr as new lines. It fails to recognize the newline at end of file which should be present as even the expectedAst suggests.

Issue is with non-main file checkstyle run. Suppression must be made at https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_non_main_files_suppressions.xml .

**Comment by ps-sp** [comment]
@rnveach ok. Also are we to not support Cr as newlines of `NewlineAtEndOfFile` check ? just asking.

**Comment by rnveach** [comment]
> are we to not support Cr as newlines of NewlineAtEndOfFile check 

It is supported, but we enforce new line to it must match system default new line. We don't want just "any" new line.
Windows is `\r\n` and Linux is `\n`.
`\r` is only allowed for Macintosh machines and is not recognized by github by default.

**Comment by ps-sp** [comment]
@rnveach my bad. Rather than checking the code I somehow just assumed it wouldn't.

**Comment by ps-sp** [comment]
[diff report 1](https://issue2329-diffreport1.netlify.com/) (manually generated the diff reports and linked them since I was inappropriately getting a `FileNotFoundException` from patch-diff-tool for checkstyle and guava projects which had also been tested for regression but are not present in report because of the exception)

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4 [hunk 4]
```diff
@@ -37,8 +37,10 @@ import java.util.*;
 
 }
 
-LEADING_ASTERISK : ( (' '|'\t') {_tokenStartCharPositionInLine == 0}? ) (' '|'\t')* '*'
-      | '*' {_tokenStartCharPositionInLine == 0}?
+LEADING_ASTERISK : ( (' '|'\t') {_tokenStartCharPositionInLine == 0
+                                 || previousTokenType == NEWLINE}? ) (' '|'\t')* '*'
+      | '*' {_tokenStartCharPositionInLine == 0 || previousTokenType == NEWLINE}?
+      | '*' {previousTokenType == NEWLINE}?
```
Comment (Vladlis) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: Isn't condition on this line covered by the previous one?

File: src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java [hunk 5]
```diff
@@ -177,6 +176,53 @@ protected static void verifyAst(String expectedTextPrintFileName, String actualJ
     }
 
     /**
+     * Verifies the Java and Javadoc Ast for the supplied java file against the supplied text file.
+     * @param expectedTextPrintFileName expected text ast tree representation.
+     * @param actualJavaFileName actual text ast tree representation.
+     * @throws Exception if exception occurs during verification.
+     */
+    protected static void verifyJavaAndJavadocAst(String expectedTextPrintFileName,
```
Comment (Vladlis) [inline src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java]: Why do we need this method?

File: src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java [hunk 6]
```diff
@@ -177,6 +176,53 @@ protected static void verifyAst(String expectedTextPrintFileName, String actualJ
...
+     * @param actualJavaFileName actual text ast tree representation.
+     * @throws Exception if exception occurs during verification.
+     */
+    protected static void verifyJavadocAst(String expectedTextPrintFileName,
+                                           String actualJavaFileName) throws Exception {
+
+        final String expectedContents = readFile(expectedTextPrintFileName);
+
+        final String actualContents = DetailNodeTreeStringPrinter.printFileAst(
+                new File(actualJavaFileName)).replaceAll("\\\\r\\\\n", "\\\\n");
```
Comment (Vladlis) [inline src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java]: Please make `"\\\\r\\\\n"` and `"\\\\n"` constants.

File: src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParseTreeTest.java [hunk 7]
```diff
@@ -36,7 +36,9 @@
 import org.junit.Assert;
 import org.junit.Test;
 
-public class JavadocParseTreeTest {
+import com.puppycrawl.tools.checkstyle.BaseCheckTestSupport;
+
+public class JavadocParseTreeTest extends BaseCheckTestSupport {
```
Comment (Vladlis) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParseTreeTest.java]: It would be better to have a separate commit for refactoring/extending of `BaseCheckTestSupport`

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4 (hunk 4 above)
Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: yes. Sorry. Missed that

File: src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java (hunk 5 above)
Comment (ps-sp) [inline src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java]: This one shall be needed in #4451. I didn't realize we won't be needing it here. I shall take it out.

File: src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParseTreeTest.java (hunk 7 above)
Comment (ps-sp) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParseTreeTest.java]: hmm. Would make the commits more readable yes. The perhaps I will clear out #3796 and #4381 first.

**Comment by romani** [comment]
@ps-sp , please rebase and reply to each point as done.

@Vladlis , please work with ps-sp to make it done and approve.

**Comment by ps-sp** [comment]
@romani Please look at [this](https://github.com/checkstyle/checkstyle/pull/4386#discussion_r122300185). I have created a PR for #3796 and will create a PR for #4381 as soon as the former one gets merged. This issue is waiting for those two.

**Comment by romani** [comment]
>  I have created a PR for #3796 

merged.

> will create a PR for #4381

please do PR, to let us review it sooner.

**Comment by ps-sp** [comment]
@romani 
> please do PR, to let us review it sooner.

Done

**Comment by romani** [comment]
@ps-sp , please rebase, you got a conflict.

@Vladlis and @sabaka , please work on this PR and approve or give student feedback.

**Comment by ps-sp** [comment]
@romani As I said this issue is blocked by #4381 . And I was away for 1 day so I couldn't reply, Sorry.

**Comment by rnveach** [comment]
@ps-sp Blocked PR and issue are done. Please rebase and fix up.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4 (hunk 4 above)
Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocLexer.g4]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java (hunk 5 above)
Comment (ps-sp) [inline src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java (hunk 6 above)
Comment (ps-sp) [inline src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParseTreeTest.java (hunk 7 above)
Comment (ps-sp) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParseTreeTest.java]: Done in PR #4486

**Comment by ps-sp** [comment]
@rnveach 

>Blocked PR and issue are done. Please rebase and fix up.

Done.

Will provide regression results as soon as I am done fixing the my other PRs.

**Comment by ps-sp** [comment]
[diff report 1](https://issue2329-diffreport1-a.netlify.com/)
